lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3771,79 +3642,1099 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+    auto *entry = jvp->createBasicBlock();
+    createEntryArguments(jvp);
+
+    // Create differential blocks and arguments.
+    // TODO: Consider visiting original blocks in pre-order (dominance) order.
+    SmallVector<SILBasicBlock *, 8> preOrderDomOrder;
+    auto &differential = getDifferential();
+    auto *origEntry = original->getEntryBlock();
+    for (auto &origBB : *original) {
+      auto *diffBB = differential.createBasicBlock();
+      diffBBMap.insert({&origBB, diffBB});
+      auto diffStructLoweredType =
+      remapType(differentialInfo.getLinearMapStructLoweredType(&origBB));
+      // If the BB is the original exit, then the differential block that we just
+      // created must be the differential function's entry. Create differential
+      // entry arguments and continue.
+      if (&origBB == origEntry) {
+        assert(diffBB->isEntry());
+        createEntryArguments(&differential);
+        auto *lastArg = diffBB->getArguments().back();
+        assert(lastArg->getType() == diffStructLoweredType);
+        differentialStructArguments[&origBB] = lastArg;
+      }
+    }
+    assert(diffBBMap.size() == 1
+           && "Can only currently handle single basic block functions");
+
+    // The differential function has type:
+    // (arg0', ..., argn', exit_diffs) -> result'.
+    auto &diffBuilder = getDifferentialBuilder();
+    auto diffParamArgs =
+        differential.getArgumentsWithoutIndirectResults().drop_back();
+    assert(diffParamArgs.size() == attr->getIndices().parameters->getCapacity());
+    auto origParamArgs = original->getArgumentsWithoutIndirectResults();
+
+    // Check if result is not varied.
+    SmallVector<SILValue, 8> origFormalResults;
+    collectAllFormalResultsInTypeOrder(*original, origFormalResults);
+    auto origResult = origFormalResults[getIndices().source];
+    // Emit warning if original result is not varied, because it will always
+    // have a zero derivative.
+    if (!activityInfo.isVaried(origResult, getIndices().parameters)) {
+      // Emit fixit if original result has a valid source location.
+      auto startLoc = origResult.getLoc().getStartSourceLoc();
+      auto endLoc = origResult.getLoc().getEndSourceLoc();
+      if (startLoc.isValid() && endLoc.isValid()) {
+        context
+        .diagnose(startLoc, diag::autodiff_nonvaried_result_fixit)
+        .fixItInsert(startLoc, "withoutDerivative(at:")
+        .fixItInsertAfter(endLoc, ")");
+      }
+    }
+
+    auto *diffEntry = getDifferential().getEntryBlock();
+    auto diffLoc = getDifferential().getLocation();
+    diffBuilder.setInsertionPoint(
+        diffEntry, getNextDifferentialLocalAllocationInsertionPoint());
+
+    for (auto index : *getIndices().parameters) {
+      auto diffParam = diffParamArgs[index];
+      auto origParam = origParamArgs[index];
+
+      if (diffParam->getType().isAddress()) {
+        // Create a local copy so that it can be written to by later adjoint
```

Comment:
Please remove unexercised code you copied from `PullbackEmitter`. In particular, this logic for address types is not applicable for tangents.